File: sdk-dart/lib/secretariat.dart
// F215-F220: Dart SDK for Secretariat
//
// Features:
// - F215: Create sdk-dart/lib/secretariat.dart file
// - F216: Define Secretariat class
// - F217: Add Socket? _socket private field
// - F218: Implement Future<String> get(String key) async method
// - F219: Connect to Unix socket at fixed path
// - F220: Send JSON-RPC request for secret.get

import 'dart:async';
import 'dart:convert';
import 'dart:io';

/// F216: Secretariat client class
///
/// Lightweight Dart SDK for communicating with the Secretariat daemon.
///
/// This SDK provides a simple async interface for retrieving secrets from
/// the local Secretariat daemon via Unix domain socket (macOS/Linux) or
/// named pipe (Windows).
///
/// Example usage:
/// ```dart
/// import 'package:secretariat/secretariat.dart';
///
/// final client = Secretariat();
/// final apiKey = await client.get('OPENAI_API_KEY');
/// print('API Key: $apiKey');
/// ```
class Secretariat {
  /// F217: Socket connection to daemon
  Socket? _socket;

  /// Path to Unix domain socket (macOS/Linux)
  static const String _defaultSocketPath = '/tmp/secretariat.sock';

  /// Named pipe path (Windows)
  static const String _defaultPipePath = r'\\.\pipe\secretariat';

  /// Custom socket/pipe path (optional)
  final String? socketPath;

  /// Request timeout duration
  final Duration timeout;

  /// Request ID counter for JSON-RPC
  int _requestId = 0;

  /// Create a new Secretariat client
  ///
  /// [socketPath] - Optional custom path to Unix socket or named pipe
  /// [timeout] - Request timeout (default: 5 seconds)
  Secretariat({
    this.socketPath,
    this.timeout = const Duration(seconds: 5),
  });

  /// F219: Get the socket path based on platform
  String get _socketPath {
    if (socketPath != null) return socketPath!;

    // Use platform-specific default path
    if (Platform.isWindows) {
      return _defaultPipePath;
    } else {
      return _defaultSocketPath;
    }
  }

  /// F219: Connect to the daemon Unix socket
  Future<void> _connect() async {
    if (_socket != null) return;

    try {
      // Connect to Unix domain socket or named pipe
      _socket = await Socket.connect(
        InternetAddress(_socketPath, type: InternetAddressType.unix),
        0,
        timeout: timeout,
      );
    } catch (e) {
      throw SecretariatException(
        'Failed to connect to Secretariat daemon at $_socketPath: $e',
      );
    }
  }

  /// F220: Send JSON-RPC request to daemon
  Future<Map<String, dynamic>> _sendRequest(
    String method,
    Map<String, dynamic> params,
  ) async {
    await _connect();

    final requestId = ++_requestId;

    // F220: Build JSON-RPC 2.0 request
    final request = {
      'jsonrpc': '2.0',
      'id': requestId,
      'method': method,
      'params': params,
    };

    final requestJson = jsonEncode(request);

    // Send request
    _socket!.write('$requestJson\n');
    await _socket!.flush();

    // Read response
    final completer = Completer<Map<String, dynamic>>();
    final responseBuffer = StringBuffer();

    _socket!.listen(
      (data) {
        responseBuffer.write(utf8.decode(data));
        final responseStr = responseBuffer.toString();

        // Check if we have a complete JSON response (ends with newline)
        if (responseStr.endsWith('\n')) {
          try {
            final response = jsonDecode(responseStr.trim()) as Map<String, dynamic>;

            // Validate response
            if (response['id'] != requestId) {
              completer.completeError(
                SecretariatException('Response ID mismatch'),
              );
              return;
            }

            if (response.containsKey('error')) {
              final error = response['error'] as Map<String, dynamic>;
              completer.completeError(
                SecretariatException(
                  error['message'] as String? ?? 'Unknown error',
                  code: error['code'] as int?,
                ),
              );
              return;
            }

            completer.complete(response);
          } catch (e) {
            completer.completeError(
              SecretariatException('Failed to parse response: $e'),
            );
          }
        }
      },
      onError: (error) {
        completer.completeError(
          SecretariatException('Socket error: $error'),
        );
      },
      onDone: () {
        if (!completer.isCompleted) {
          completer.completeError(
            SecretariatException('Connection closed unexpectedly'),
          );
        }
      },
    );

    return completer.future.timeout(
      timeout,
      onTimeout: () {
        throw SecretariatException('Request timed out');
      },
    );
  }

  /// F218: Get secret value by key
  ///
  /// Retrieves the decrypted value of a secret from the daemon.
  ///
  /// [key] - Secret name/key (e.g., 'OPENAI_API_KEY')
  ///
  /// Returns the secret value as a String.
  ///
  /// Throws [SecretariatException] if:
  /// - Cannot connect to daemon
  /// - Secret not found
  /// - Permission denied
  /// - Network/communication error
  ///
  /// Example:
  /// ```dart
  /// final client = Secretariat();
  /// try {
  ///   final apiKey = await client.get('OPENAI_API_KEY');
  ///   print('Retrieved: $apiKey');
  /// } catch (e) {
  ///   print('Error: $e');
  /// }
  /// ```
  Future<String> get(String key) async {
    // F220: Send secret.get JSON-RPC request
    final response = await _sendRequest('secret.get', {'key': key});

    // Extract result from response
    if (!response.containsKey('result')) {
      throw SecretariatException('Invalid response: missing result');
    }

    final result = response['result'] as Map<String, dynamic>;

    if (!result.containsKey('value')) {
      throw SecretariatException('Invalid response: missing value');
    }

    return result['value'] as String;
  }

  /// Get multiple secrets at once
  ///
  /// [keys] - List of secret names to retrieve
  ///
  /// Returns a Map of key -> value pairs.
  ///
  /// Example:
  /// ```dart
  /// final secrets = await client.getMany([
  ///   'OPENAI_API_KEY',
  ///   'DATABASE_URL',
  /// ]);
  /// print(secrets['OPENAI_API_KEY']);
  /// ```
  Future<Map<String, String>> getMany(List<String> keys) async {
    final results = <String, String>{};

    for (final key in keys) {
      try {
        results[key] = await get(key);
      } catch (e) {
        // Continue on error, return partial results
        rethrow;
      }
    }

    return results;
  }

  /// List all available secret names
  ///
  /// Returns a list of secret names (not values).
  ///
  /// Example:
  /// ```dart
  /// final names = await client.list();
  /// print('Available secrets: $names');
  /// ```
  Future<List<String>> list() async {
    final response = await _sendRequest('secret.list', {});

    if (!response.containsKey('result')) {
      throw SecretariatException('Invalid response: missing result');
    }

    final result = response['result'] as Map<String, dynamic>;

    if (!result.containsKey('secrets')) {
      throw SecretariatException('Invalid response: missing secrets');
    }

    final secrets = result['secrets'] as List<dynamic>;
    return secrets.map((s) => s as String).toList();
  }

  /// Close the connection to the daemon
  ///
  /// Call this when you're done using the client to free resources.
  Future<void> close() async {
    await _socket?.close();
    _socket = null;
  }
}

/// Exception thrown by Secretariat SDK
class SecretariatException implements Exception {
  /// Error message
  final String message;

  /// Optional error code
  final int? code;

  SecretariatException(this.message, {this.code});

  @override
  String toString() {
    if (code != null) {
      return 'SecretariatException($code): $message';
    }
    return 'SecretariatException: $message';
  }
}

File: sdk-go/secretariat.go
// Package secretariat provides a Go client for the Secretariat secrets manager daemon.
//
// Secretariat is a local-first secrets manager that stores API keys, tokens, and
// other secrets securely. This SDK communicates with the local Secretariat daemon
// via Unix domain socket.
//
// Example usage:
//
//	client, err := secretariat.New()
//	if err != nil {
//	    log.Fatal(err)
//	}
//	defer client.Close()
//
//	apiKey, err := client.Get("OPENAI_API_KEY")
//	if err != nil {
//	    log.Fatal(err)
//	}
//	fmt.Println("API Key:", apiKey)
package secretariat

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"os"
	"path/filepath"
	"runtime"
	"sync"
	"time"
)

// Default socket paths by platform
const (
	defaultTimeoutSeconds = 5
)

// Error types
var (
	// ErrNotConnected is returned when the client is not connected to the daemon.
	ErrNotConnected = errors.New("not connected to daemon")

	// ErrNotFound is returned when a secret is not found.
	ErrNotFound = errors.New("secret not found")

	// ErrPermissionDenied is returned when access to a secret is denied.
	ErrPermissionDenied = errors.New("permission denied")

	// ErrVaultLocked is returned when the vault is locked.
	ErrVaultLocked = errors.New("vault is locked")

	// ErrTimeout is returned when a request times out.
	ErrTimeout = errors.New("request timed out")
)

// SecretariatError represents an error from the Secretariat daemon.
type SecretariatError struct {
	Message string
	Code    int
}

func (e *SecretariatError) Error() string {
	if e.Code != 0 {
		return fmt.Sprintf("secretariat error (%d): %s", e.Code, e.Message)
	}
	return fmt.Sprintf("secretariat error: %s", e.Message)
}

// jsonRPCRequest is the JSON-RPC 2.0 request format.
type jsonRPCRequest struct {
	JSONRPC string                 `json:"jsonrpc"`
	ID      int                    `json:"id"`
	Method  string                 `json:"method"`
	Params  map[string]interface{} `json:"params"`
}

// jsonRPCResponse is the JSON-RPC 2.0 response format.
type jsonRPCResponse struct {
	JSONRPC string                 `json:"jsonrpc"`
	ID      int                    `json:"id"`
	Result  map[string]interface{} `json:"result,omitempty"`
	Error   *jsonRPCError          `json:"error,omitempty"`
}

type jsonRPCError struct {
	Code    int    `json:"code"`
	Message string `json:"message"`
}

// SecretInfo contains metadata about a secret.
type SecretInfo struct {
	Name        string `json:"name"`
	Provider    string `json:"provider,omitempty"`
	Environment string `json:"environment,omitempty"`
	CreatedAt   string `json:"created_at,omitempty"`
	UpdatedAt   string `json:"updated_at,omitempty"`
}

// Client is a Secretariat daemon client.
type Client struct {
	socketPath string
	timeout    time.Duration
	conn       net.Conn
	reader     *bufio.Reader
	requestID  int
	mu         sync.Mutex
}

// Option is a functional option for configuring the Client.
type Option func(*Client)

// WithSocketPath sets a custom socket path.
func WithSocketPath(path string) Option {
	return func(c *Client) {
		c.socketPath = path
	}
}

// WithTimeout sets the request timeout.
func WithTimeout(d time.Duration) Option {
	return func(c *Client) {
		c.timeout = d
	}
}

// New creates a new Secretariat client.
// It automatically detects the platform and uses the appropriate socket path.
func New(opts ...Option) (*Client, error) {
	c := &Client{
		timeout: defaultTimeoutSeconds * time.Second,
	}

	for _, opt := range opts {
		opt(c)
	}

	// Determine default socket path if not set
	if c.socketPath == "" {
		c.socketPath = defaultSocketPath()
	}

	// Connect to daemon
	if err := c.connect(); err != nil {
		return nil, fmt.Errorf("failed to connect to daemon: %w", err)
	}

	return c, nil
}

// defaultSocketPath returns the default socket path for the current platform.
func defaultSocketPath() string {
	if env := os.Getenv("SECRETARIAT_SOCKET_PATH"); env != "" {
		return env
	}
	if env := os.Getenv("SECRETARIAT_SOCKET"); env != "" {
		return env
	}

	switch runtime.GOOS {
	case "darwin": // macOS
		home, _ := os.UserHomeDir()
		return filepath.Join(home, "Library", "Application Support", "Secretariat", "secretariat.sock")
	case "linux":
		home, _ := os.UserHomeDir()
		return filepath.Join(home, ".local", "share", "secretariat", "secretariat.sock")
	case "windows":
		return `\\.\pipe\secretariat`
	default:
		home, _ := os.UserHomeDir()
		return filepath.Join(home, ".local", "share", "secretariat", "secretariat.sock")
	}
}

// connect establishes connection to the daemon.
func (c *Client) connect() error {
	conn, err := net.DialTimeout("unix", c.socketPath, c.timeout)
	if err != nil {
		return fmt.Errorf("cannot connect to %s: %w", c.socketPath, err)
	}
	c.conn = conn
	c.reader = bufio.NewReader(conn)
	return nil
}

// Close closes the connection to the daemon.
func (c *Client) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn != nil {
		err := c.conn.Close()
		c.conn = nil
		c.reader = nil
		return err
	}
	return nil
}

// request sends a JSON-RPC request and returns the response.
func (c *Client) request(method string, params map[string]interface{}) (map[string]interface{}, error) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.conn == nil {
		return nil, ErrNotConnected
	}

	// Increment request ID
	c.requestID++
	id := c.requestID

	// Build request
	req := jsonRPCRequest{
		JSONRPC: "2.0",
		ID:      id,
		Method:  method,
		Params:  params,
	}

	// Encode and send
	data, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to encode request: %w", err)
	}

	// Set deadline
	if err := c.conn.SetDeadline(time.Now().Add(c.timeout)); err != nil {
		return nil, fmt.Errorf("failed to set deadline: %w", err)
	}

	// Send with newline terminator
	if _, err := c.conn.Write(append(data, '\n')); err != nil {
		return nil, fmt.Errorf("failed to send request: %w", err)
	}

	// Read response
	line, err := c.reader.ReadBytes('\n')
	if err != nil {
		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
			return nil, ErrTimeout
		}
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	// Parse response
	var resp jsonRPCResponse
	if err := json.Unmarshal(line, &resp); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	// Check ID
	if resp.ID != id {
		return nil, fmt.Errorf("response ID mismatch: expected %d, got %d", id, resp.ID)
	}

	// Check for error
	if resp.Error != nil {
		return nil, &SecretariatError{
			Message: resp.Error.Message,
			Code:    resp.Error.Code,
		}
	}

	return resp.Result, nil
}

// Get retrieves a secret value by name.
//
// Example:
//
//	apiKey, err := client.Get("OPENAI_API_KEY")
func (c *Client) Get(name string) (string, error) {
	return c.GetWithOptions(name, GetOptions{})
}

// GetOptions configures a Get request.
type GetOptions struct {
	// AppID identifies the requesting application for permission checks.
	AppID string

	// Environment specifies which environment to get the secret from.
	Environment string
}

// GetWithOptions retrieves a secret with additional options.
func (c *Client) GetWithOptions(name string, opts GetOptions) (string, error) {
	params := map[string]interface{}{
		"name": name,
	}
	appID := opts.AppID
	if appID == "" {
		appID = "go-sdk"
	}
	params["app_id"] = appID
	if opts.Environment != "" {
		params["environment"] = opts.Environment
	}

	result, err := c.request("secret.get", params)
	if err != nil {
		return "", err
	}

	value, ok := result["value"].(string)
	if !ok {
		return "", fmt.Errorf("invalid response: missing value")
	}

	return value, nil
}

// GetMany retrieves multiple secrets at once.
//
// Example:
//
//	secrets, err := client.GetMany([]string{"OPENAI_API_KEY", "DATABASE_URL"})
func (c *Client) GetMany(names []string) (map[string]string, error) {
	results := make(map[string]string, len(names))
	for _, name := range names {
		value, err := c.Get(name)
		if err != nil {
			return nil, fmt.Errorf("failed to get %s: %w", name, err)
		}
		results[name] = value
	}
	return results, nil
}

// List returns all available secret names.
//
// Example:
//
//	names, err := client.List()
func (c *Client) List() ([]SecretInfo, error) {
	result, err := c.request("secret.list", map[string]interface{}{})
	if err != nil {
		return nil, err
	}

	secretsRaw, ok := result["secrets"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("invalid response: missing secrets")
	}

	secrets := make([]SecretInfo, 0, len(secretsRaw))
	for _, s := range secretsRaw {
		switch v := s.(type) {
		case map[string]interface{}:
			info := SecretInfo{}
			if name, ok := v["name"].(string); ok {
				info.Name = name
			}
			if provider, ok := v["provider"].(string); ok {
				info.Provider = provider
			}
			if env, ok := v["environment"].(string); ok {
				info.Environment = env
			}
			secrets = append(secrets, info)
		case string:
			secrets = append(secrets, SecretInfo{Name: v})
		}
	}

	return secrets, nil
}

// ListNames returns just the secret names.
func (c *Client) ListNames() ([]string, error) {
	secrets, err := c.List()
	if err != nil {
		return nil, err
	}

	names := make([]string, len(secrets))
	for i, s := range secrets {
		names[i] = s.Name
	}
	return names, nil
}

// Set creates or updates a secret.
//
// Example:
//
//	err := client.Set("API_KEY", "sk-123456789")
func (c *Client) Set(name, value string) error {
	return c.SetWithOptions(name, value, SetOptions{})
}

// SetOptions configures a Set request.
type SetOptions struct {
	// Environment specifies which environment to store the secret in.
	Environment string

	// Provider identifies the secret provider (e.g., "openai", "stripe").
	Provider string
}

// SetWithOptions creates or updates a secret with additional options.
func (c *Client) SetWithOptions(name, value string, opts SetOptions) error {
	params := map[string]interface{}{
		"name":  name,
		"value": value,
	}
	if opts.Environment != "" {
		params["environment"] = opts.Environment
	}
	if opts.Provider != "" {
		params["provider"] = opts.Provider
	}

	_, err := c.request("secret.set", params)
	return err
}

// Delete removes a secret.
//
// Example:
//
//	err := client.Delete("OLD_API_KEY")
func (c *Client) Delete(name string) error {
	_, err := c.request("secret.delete", map[string]interface{}{
		"name": name,
	})
	return err
}

// Ping checks if the daemon is responding.
func (c *Client) Ping() error {
	_, err := c.request("health.check", map[string]interface{}{})
	return err
}

// MustGet retrieves a secret or panics on error.
// Useful for initialization code where failure is fatal.
//
// Example:
//
//	apiKey := client.MustGet("OPENAI_API_KEY")
func (c *Client) MustGet(name string) string {
	value, err := c.Get(name)
	if err != nil {
		panic(fmt.Sprintf("secretariat: failed to get %s: %v", name, err))
	}
	return value
}

// Get is a convenience function that creates a client, gets a secret, and closes.
//
// Example:
//
//	apiKey, err := secretariat.Get("OPENAI_API_KEY")
func Get(name string) (string, error) {
	client, err := New()
	if err != nil {
		return "", err
	}
	defer client.Close()

	return client.Get(name)
}

// MustGet is a convenience function that gets a secret or panics.
//
// Example:
//
//	apiKey := secretariat.MustGet("OPENAI_API_KEY")
func MustGet(name string) string {
	value, err := Get(name)
	if err != nil {
		panic(fmt.Sprintf("secretariat: failed to get %s: %v", name, err))
	}
	return value
}

// GetOrEnv gets a secret with environment variable fallback.
//
// It first tries to get the secret from the daemon. If that fails,
// it falls back to the environment variable.
//
// Example:
//
//	apiKey := secretariat.GetOrEnv("OPENAI_API_KEY", "")
func GetOrEnv(name, envVar string) string {
	if envVar == "" {
		envVar = name
	}

	value, err := Get(name)
	if err == nil {
		return value
	}

	// Fall back to environment variable
	return os.Getenv(envVar)
}
